pkg/cloudprovider/aws/instancetypes.go

```diff
@@ -0,0 +1,226 @@
+		var err error
+		locations, err = p.vpc.GetZones(ctx, clusterName)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// locations could be zone names and zone ids, so normalize them to all be zone names
+	locations, err := p.normalizeZones(ctx, locations)
+	if err != nil {
+		return nil, err
+	}
+
+	zoneToInstanceTypeInfo := map[string][]*ec2.InstanceTypeInfo{}
+	for _, zone := range locations {
+		if instanceTypes, ok := p.instanceTypeCache.Get(zone); ok {
+			zoneToInstanceTypeInfo[zone] = instanceTypes.([]*ec2.InstanceTypeInfo)
+			continue
+		}
+
+		// populate the cache by zonal keys
+		instanceTypes, err := p.getAllInstanceTypes(ctx)
+		if err != nil {
+			return nil, err
+		}
+		instanceTypes, err = p.filterByLocation(ctx, instanceTypes, zone)
+		if err != nil {
+			return nil, err
+		}
+		p.instanceTypeCache.SetDefault(zone, instanceTypes)
+		zoneToInstanceTypeInfo[zone] = instanceTypes
+	}
+
+	for zone, instanceTypes := range zoneToInstanceTypeInfo {
+		zoneToInstanceTypeInfo[zone] = p.filterFrom(instanceTypes, constraints)
+	}
+	return zoneToInstanceTypeInfo, nil
+}
+
+// UniqueInstanceTypesFrom returns a unique slice of InstanceTypeInfo structs from the values of the instancePools passed in
+func (p *InstanceTypeProvider) UniqueInstanceTypesFrom(instancePools map[string][]*ec2.InstanceTypeInfo) []*ec2.InstanceTypeInfo {
+	result := []*ec2.InstanceTypeInfo{}
+	uniqueInstanceTypes := map[string]bool{}
+	for _, instanceTypes := range instancePools {
+		for _, instanceType := range instanceTypes {
+			if _, ok := uniqueInstanceTypes[*instanceType.InstanceType]; !ok {
+				uniqueInstanceTypes[*instanceType.InstanceType] = true
+				result = append(result, instanceType)
+			}
+		}
+	}
+	return result
+}
+
+// InstanceTypesPerZoneFrom returns a mapping of zone to InstanceTypeInfo based on the provided slice of instance types names and mapping of all zones to InstanceTypeInfo
+func (p *InstanceTypeProvider) InstanceTypesPerZoneFrom(instanceTypeNames []string, instancePools map[string][]*ec2.InstanceTypeInfo) map[string][]*ec2.InstanceTypeInfo {
+	result := map[string][]*ec2.InstanceTypeInfo{}
+	for zone, instanceTypes := range instancePools {
+		for _, instanceTypeName := range instanceTypeNames {
+			for _, instanceType := range instanceTypes {
+				if instanceType != nil && *instanceType.InstanceType == instanceTypeName {
+					result[zone] = append(result[zone], instanceType)
+				}
+			}
+		}
+	}
+	return result
+}
+
+// getAllInstanceTypes retrieves all instance types from the ec2 DescribeInstanceTypes API using some opinionated filters
+func (p *InstanceTypeProvider) getAllInstanceTypes(ctx context.Context) ([]*ec2.InstanceTypeInfo, error) {
+	instanceTypes := []*ec2.InstanceTypeInfo{}
+	describeInstanceTypesInput := &ec2.DescribeInstanceTypesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("supported-virtualization-type"),
+				Values: []*string{aws.String("hvm")},
+			},
+			{
+				Name:   aws.String("bare-metal"),
```

Comment:
Why are we excluding metal types?

Reply:
Metal definitely makes sense in some cases, it all depends on what people are doing with their cluster, and we don't want to exclude it. That said, we probably need to test metal before enabling it so I could see excluding it for now (just add a TODO that we'll reenable after testing).

Reply:
Maybe we exclude it from default instance types but allow it to be specified directly

Reply:
Does that require any further code change or could someone do it with labels or something?

Reply:
That's a good point on being able to specify it directly right now...  I'll move these filters into the custom filter logic so that all instance types are still available for direct override by customers in the instance types config or via node selector.